File: sdk/ios/Sources/Tool.swift
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.

import Foundation

/// Type-erased Tool facade — the view the `Agent` sees.
///
/// `Tool` refines this with typed `Args` / `Ret`. Type erasure lets
/// heterogeneous tools live in the same `[any ErasedTool]` collection
/// without forcing the caller to match associated types.
public protocol ErasedTool: Sendable {
    var name: String { get }
    var description: String { get }
    var argsSchema: JsonSchema { get }
    func toDeclaration() -> ToolDeclaration

    /// Invoke via raw JSON arguments. Default implementation on `Tool`
    /// decodes → invokes → encodes; custom conformers can override.
    func invokeRaw(arguments raw: String, ctx: ToolContext) async throws -> String
}

public extension ErasedTool {
    func toDeclaration() -> ToolDeclaration {
        ToolDeclaration(name: name, description: description, parameters: argsSchema)
    }
}

/// A typed function that an LLM / `Agent` can invoke during a conversation.
///
/// Serializes to the **OpenAI / Anthropic / Gemini function-calling format**
/// via `toDeclaration()`, so a `Tool` written in Swift is consumed as-is
/// by every modern LLM runtime.
public protocol Tool<Args, Ret>: ErasedTool {
    associatedtype Args
    associatedtype Ret

    /// Execute the tool. Called after `argsFromJson(_:)` decodes the
    /// LLM's tool_call arguments into typed `Args`.
    func invoke(args: Args, ctx: ToolContext) async throws -> Ret

    /// Decode the LLM's tool_call `arguments` JSON into typed `Args`.
    func argsFromJson(_ raw: String) throws -> Args

    /// Encode the return value back to a JSON string that the SDK
    /// appends as the content of a `role == .tool` response message.
    func returnToJson(_ value: Ret) -> String
}

public extension Tool {
    /// Default erased invocation: decode → invoke → encode.
    func invokeRaw(arguments raw: String, ctx: ToolContext) async throws -> String {
        let args = try argsFromJson(raw)
        let result = try await invoke(args: args, ctx: ctx)
        return returnToJson(result)
    }
}

/// Wire-format tool declaration. Serialize to JSON and paste directly
/// into an OpenAI `tools: [...]` array — no adapter needed.
public struct ToolDeclaration: Sendable {
    public let name: String
    public let description: String
    public let parameters: JsonSchema

    public init(name: String, description: String, parameters: JsonSchema) {
        self.name = name
        self.description = description
        self.parameters = parameters
    }
}

/// Execution context available inside `Tool.invoke`.
public struct ToolContext: @unchecked Sendable {
    /// Threading / concurrency settings inherited from the agent.
    public let execution: ExecutionPolicy
    /// Named context stores the tool can read from or write to.
    public let stores: [String: any ContextStoreBox]
    /// Reserved for multi-agent pub/sub (Layer 3+).
    public let publish: (@Sendable (String, String) async -> Void)?

    public init(
        execution: ExecutionPolicy,
        stores: [String: any ContextStoreBox] = [:],
        publish: (@Sendable (String, String) async -> Void)? = nil
    ) {
        self.execution = execution
        self.stores = stores
        self.publish = publish
    }
}

/// Type-erasing wrapper so `ToolContext` can carry heterogeneous
/// `ContextStore<T>` values. Concrete stores conform automatically.
public protocol ContextStoreBox: AnyObject, Sendable {
    var name: String { get }
}

// ── JsonSchema ───────────────────────────────────────────────────────────

/// Minimal JSON Schema representation — enough to describe tool
/// parameter shapes without pulling a full JSON Schema implementation.
public indirect enum JsonSchema: Sendable {
    /// A parameters object with named fields.
    case object(properties: [(String, JsonSchema)], required: [String], description: String?)
    /// A scalar: "string" / "integer" / "number" / "boolean".
    case primitive(type: String, description: String?, enumValues: [String]?, minimum: Double?, maximum: Double?)
    /// An array of a single item shape.
    case array(items: JsonSchema, description: String?)
    /// Pass-through — the JSON string was already built in Dart / TS
    /// and the native layer only needs to forward it verbatim into
    /// the LLM prompt. Used by the Flutter + React Native bridges.
    case raw(json: String)

    /// Emit this schema as a JSON string (OpenAI-compatible).
    public func serialize() -> String {
        switch self {
        case .object(let props, let required, let desc):
            var s = "{\"type\":\"object\""
            if let desc = desc { s += ",\"description\":\"\(JsonSchema.escape(desc))\"" }
            s += ",\"properties\":{"
            for (i, (k, v)) in props.enumerated() {
                if i > 0 { s += "," }
                s += "\"\(JsonSchema.escape(k))\":\(v.serialize())"
            }
            s += "}"
            if !required.isEmpty {
                s += ",\"required\":["
                for (i, r) in required.enumerated() {
                    if i > 0 { s += "," }
                    s += "\"\(JsonSchema.escape(r))\""
                }
                s += "]"
            }
            return s + "}"
        case .primitive(let type, let desc, let enumVals, let minV, let maxV):
            precondition(["string", "integer", "number", "boolean"].contains(type),
                "unsupported primitive type: \(type)")
            var s = "{\"type\":\"\(type)\""
            if let desc = desc { s += ",\"description\":\"\(JsonSchema.escape(desc))\"" }
            if let e = enumVals, !e.isEmpty {
                s += ",\"enum\":["
                for (i, v) in e.enumerated() {
                    if i > 0 { s += "," }
                    s += "\"\(JsonSchema.escape(v))\""
                }
                s += "]"
            }
            if let m = minV { s += ",\"minimum\":\(m)" }
            if let m = maxV { s += ",\"maximum\":\(m)" }
            return s + "}"
        case .array(let items, let desc):
            var s = "{\"type\":\"array\",\"items\":\(items.serialize())"
            if let desc = desc { s += ",\"description\":\"\(JsonSchema.escape(desc))\"" }
            return s + "}"
        case .raw(let json):
            return json
        }
    }

    private static func escape(_ s: String) -> String {
        var out = ""
        for c in s {
            switch c {
            case "\"":  out += "\\\""
            case "\\":  out += "\\\\"
            case "\n":  out += "\\n"
            case "\r":  out += "\\r"
            case "\t":  out += "\\t"
            default:    out.append(c)
            }
        }
        return out
    }
}

// ── Builder DSL for JsonSchema.object ───────────────────────────────────

/// Build a JsonSchema object with a fluent API.
///
/// ```swift
/// let schema = jsonSchemaObject(description: "Weather query") {
///     $0.property("city", type: "string", required: true,
///                 description: "City name, e.g. 'Lima'")
///     $0.property("unit", type: "string",
///                 description: "C or F", enum: ["C", "F"])
/// }
/// ```
public func jsonSchemaObject(
    description: String? = nil,
    _ build: (inout JsonSchemaObjectBuilder) -> Void
) -> JsonSchema {
    var b = JsonSchemaObjectBuilder(description: description)
    build(&b)
    return b.build()
}

public struct JsonSchemaObjectBuilder {
    private let description: String?
    private var properties: [(String, JsonSchema)] = []
    private var required: [String] = []

    fileprivate init(description: String?) { self.description = description }

    public mutating func property(
        _ name: String,
        type: String,
        description: String? = nil,
        required: Bool = false,
        `enum` enumValues: [String]? = nil,
        minimum: Double? = nil,
        maximum: Double? = nil
    ) {
        let s = JsonSchema.primitive(
            type: type,
            description: description,
            enumValues: enumValues,
            minimum: minimum,
            maximum: maximum
        )
        properties.append((name, s))
        if required { self.required.append(name) }
    }

    public mutating func property(_ name: String, schema: JsonSchema, required: Bool = false) {
        properties.append((name, schema))
        if required { self.required.append(name) }
    }

    fileprivate func build() -> JsonSchema {
        .object(properties: properties, required: required, description: description)
    }
}

File: sdk/android/src/main/java/dev/dazzle/sdk/Tool.kt
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.

package dev.dazzle.sdk

/**
 * A typed function that an LLM / [Agent] can invoke during a conversation.
 *
 * Serializes to the **OpenAI / Anthropic / Gemini function-calling format**
 * via [toDeclaration], so a `Tool` written in Kotlin is consumed as-is by
 * every modern LLM runtime: LiteRT-LM, llama.cpp, Ollama, vLLM, or a
 * hosted API.
 *
 * ## Shape
 *
 * `Args` is whatever typed input the tool needs (typically a `data class`).
 * `Ret` is the value returned to the caller. Both are bridged via
 * [argsFromJson] / [returnToJson] — the dev controls the exact JSON layout
 * so there's no hidden encoding magic.
 *
 * ## Example
 *
 * ```kotlin
 * data class WeatherQuery(val city: String, val unit: String = "C")
 * data class WeatherReport(val tempC: Double, val condition: String)
 *
 * val weatherTool = object : Tool<WeatherQuery, WeatherReport> {
 *     override val name = "weather.get"
 *     override val description = "Return the current weather for a city."
 *     override val argsSchema = jsonSchemaObject {
 *         property("city", type = "string", required = true,
 *                  description = "City name, e.g. 'Lima'")
 *         property("unit", type = "string", required = false,
 *                  description = "Temperature unit: C or F")
 *     }
 *     override suspend fun invoke(args: WeatherQuery, ctx: ToolContext): WeatherReport =
 *         myWeatherApi.fetch(args.city, args.unit)
 *
 *     override fun argsFromJson(raw: String): WeatherQuery = parseJson(raw)
 *     override fun returnToJson(value: WeatherReport): String = toJson(value)
 * }
 * ```
 */
interface Tool<Args, Ret> {
    /** Canonical identifier the LLM references in tool_calls. Must be
     *  unique within an [Agent]'s tool set; convention is `domain.verb`. */
    val name: String

    /** Natural-language description the LLM reads to decide *when* to
     *  invoke this tool. Keep it concise (≤200 chars) and action-oriented. */
    val description: String

    /** JSON Schema describing the shape of [Args]. Used by the model to
     *  produce well-formed tool_calls and by the caller to validate before
     *  dispatch. See [jsonSchemaObject] for a minimal builder. */
    val argsSchema: JsonSchema

    /** Execute the tool. Called after [argsFromJson] decodes the LLM's
     *  tool_call arguments into typed [Args]. The [ctx] grants access to
     *  the agent's stores and execution policy. */
    suspend fun invoke(args: Args, ctx: ToolContext): Ret

    /** Decode a tool_call's `arguments` JSON (which the LLM produced per
     *  [argsSchema]) into typed [Args]. Throws if the JSON is malformed. */
    fun argsFromJson(raw: String): Args

    /** Encode the return value back to a JSON string that the SDK appends
     *  as the `content` of a `Role.tool` response message. */
    fun returnToJson(value: Ret): String

    /** Serialize to the shape every LLM tool-calling API expects:
     *  `{ "name": ..., "description": ..., "parameters": <JSON Schema> }`. */
    fun toDeclaration(): ToolDeclaration =
        ToolDeclaration(name = name, description = description, parameters = argsSchema)
}

/**
 * Wire-format tool declaration. Serialize to JSON and paste directly into
 * an OpenAI `tools: [...]` array — no adapter needed.
 *
 * ```json
 * { "name": "weather.get",
 *   "description": "Return the current weather for a city.",
 *   "parameters": { "type": "object", "properties": {...}, "required": [...] } }
 * ```
 */
data class ToolDeclaration(
    val name: String,
    val description: String,
    val parameters: JsonSchema,
)

/**
 * Execution context available inside [Tool.invoke]. Gives the tool access
 * to the agent's shared context stores and execution policy without
 * exposing the whole Dazzle surface.
 */
class ToolContext(
    /** Threading / concurrency settings inherited from the agent. */
    val execution: ExecutionPolicy,

    /** Named context stores the tool can read from or write to. Keys are
     *  whatever the agent's builder registered. Typical keys: `"memory"`
     *  (chat turns), `"knowledge"` (RAG documents), `"sensors"` (domain). */
    val stores: Map<String, ContextStore<*>>,

    /** Reserved for multi-agent pub/sub. When null, publishing is a no-op.
     *  Typed shape lands in a future `Channel<T>` follow-up. */
    val publish: (suspend (channel: String, message: String) -> Unit)? = null,
)

/**
 * Minimal JSON Schema representation — just enough to describe tool
 * parameter shapes without pulling a full JSON Schema implementation.
 *
 * Three levels suffice for 95% of tools:
 *
 *   - [ObjectSchema]: a parameters object with named fields
 *   - [PrimitiveSchema]: `string`, `integer`, `number`, `boolean`
 *   - [ArraySchema]: lists of a single item shape
 *
 * Serialized to the JSON string expected by OpenAI / Anthropic via
 * [serialize]. Kept intentionally loose — devs who need `anyOf`, `$ref`,
 * etc. can subclass and emit their own JSON.
 */
sealed class JsonSchema {

    /** Emit this schema as a JSON string (OpenAI-compatible). */
    abstract fun serialize(): String

    data class ObjectSchema(
        val properties: Map<String, JsonSchema>,
        val required: List<String> = emptyList(),
        val description: String? = null,
    ) : JsonSchema() {
        override fun serialize(): String = buildString {
            append("{\"type\":\"object\",")
            description?.let { append("\"description\":\"${escapeJson(it)}\",") }
            append("\"properties\":{")
            properties.entries.forEachIndexed { i, (k, v) ->
                if (i > 0) append(',')
                append('"'); append(escapeJson(k)); append("\":"); append(v.serialize())
            }
            append('}')
            if (required.isNotEmpty()) {
                append(",\"required\":[")
                required.forEachIndexed { i, r ->
                    if (i > 0) append(',')
                    append('"'); append(escapeJson(r)); append('"')
                }
                append(']')
            }
            append('}')
        }
    }

    data class PrimitiveSchema(
        val type: String,                      // "string" / "integer" / "number" / "boolean"
        val description: String? = null,
        val enum: List<String>? = null,
        val minimum: Double? = null,
        val maximum: Double? = null,
    ) : JsonSchema() {
        init {
            require(type in setOf("string", "integer", "number", "boolean")) {
                "unsupported primitive type: $type"
            }
        }
        override fun serialize(): String = buildString {
            append("{\"type\":\"").append(type).append('"')
            description?.let { append(",\"description\":\"").append(escapeJson(it)).append('"') }
            enum?.let {
                append(",\"enum\":[")
                it.forEachIndexed { i, e ->
                    if (i > 0) append(',')
                    append('"'); append(escapeJson(e)); append('"')
                }
                append(']')
            }
            minimum?.let { append(",\"minimum\":").append(it) }
            maximum?.let { append(",\"maximum\":").append(it) }
            append('}')
        }
    }

    data class ArraySchema(
        val items: JsonSchema,
        val description: String? = null,
    ) : JsonSchema() {
        override fun serialize(): String = buildString {
            append("{\"type\":\"array\",\"items\":").append(items.serialize())
            description?.let { append(",\"description\":\"").append(escapeJson(it)).append('"') }
            append('}')
        }
    }

    /**
     * Pass-through schema that already ships as a JSON string — used
     * by the Flutter / React Native bridges where the consumer builds
     * the schema in Dart / TypeScript and the native layer only needs
     * to forward it verbatim into the LLM prompt. Consumers of the
     * Kotlin SDK directly should stick to the typed variants above.
     */
    data class RawSchema(val json: String) : JsonSchema() {
        override fun serialize(): String = json
    }

    companion object {
        private fun escapeJson(s: String): String = buildString(s.length) {
            for (c in s) when (c) {
                '"'  -> append("\\\"")
                '\\' -> append("\\\\")
                '\n' -> append("\\n")
                '\r' -> append("\\r")
                '\t' -> append("\\t")
                else -> append(c)
            }
        }
    }
}

// ── Minimal DSL for building an ObjectSchema — no third-party deps ──────

/**
 * Build an [JsonSchema.ObjectSchema] with a fluent-ish API.
 *
 * ```kotlin
 * val schema = jsonSchemaObject(description = "Weather query") {
 *     property("city", type = "string", required = true, description = "city name")
 *     property("unit", type = "string", description = "C or F", enum = listOf("C", "F"))
 * }
 * ```
 */
fun jsonSchemaObject(
    description: String? = null,
    build: JsonSchemaObjectBuilder.() -> Unit,
): JsonSchema.ObjectSchema = JsonSchemaObjectBuilder(description).apply(build).build()

class JsonSchemaObjectBuilder internal constructor(private val description: String?) {
    private val props = linkedMapOf<String, JsonSchema>()
    private val reqList = mutableListOf<String>()

    @JvmOverloads
    fun property(
        name: String,
        type: String,
        description: String? = null,
        required: Boolean = false,
        enum: List<String>? = null,
        minimum: Double? = null,
        maximum: Double? = null,
    ) {
        props[name] = JsonSchema.PrimitiveSchema(
            type = type,
            description = description,
            enum = enum,
            minimum = minimum,
            maximum = maximum,
        )
        if (required) reqList += name
    }

    fun property(name: String, schema: JsonSchema, required: Boolean = false) {
        props[name] = schema
        if (required) reqList += name
    }

    internal fun build() = JsonSchema.ObjectSchema(
        properties = props,
        required = reqList.toList(),
        description = description,
    )
}
